pyxdf/pyxdf.py

```diff
@@ -256,13 +256,24 @@ def load_xdf(filename,
             # read [Tag]
             tag = struct.unpack('<H', f.read(2))[0]
             log_str = ' Read tag: {} at {} bytes, length={}'.format(tag, f.tell(), chunklen)
+            StreamId = None
             if tag in [2, 3, 4, 6]:
-                StreamId = struct.unpack('<I', f.read(4))[0]
-                log_str += ', StreamId={}'.format(StreamId)
-            else:
-                StreamId = None
-
-            logger.debug(log_str)
+                _streamid = f.read(4)
+                try:
+                    StreamId = struct.unpack("<I", _streamid)[0]
+                    log_str += ", StreamId={}".format(StreamId)
+                    logger.debug(log_str)
```

Comment:
These lines will not be reached if a `struct.error` is raised in the previous statement. Therefore, I'd move these lines to an `else` block.

Reply:
I guess an alternative using an else-block could be:

``` python
StreamId = None
if tag in [2, 3, 4, 6]:
    _streamid = f.read(4)
    try:
        StreamId = struct.unpack("<I", _streamid)[0]
    except struct.error:
        pass
        
    if StreamId is None:            
        # If the StreamId could not be parsed correctly, it will be
        # None. 
        # we scan forward to next (hopefully) valid block in a bid
        # to load as much of the file as possible. We also need to
        # continue, because otherwise we might end up in one
        # the tag-specific branches which expect a valid StreamId
        log_str += ", but StreamId is corrupt, scanning forward to next boundary chunk."
        logger.error(log_str)
        _scan_forward(f)
        continue
    else:
        # the StreamId was successfully unpacked
        log_str += ", StreamId={}".format(StreamId)
        logger.debug(log_str)


```
I didn't benchmark or inspect the bytecode, but this simply adds an if-clause and is therefore very likely the more expensive solution. I also have the impression it separates the logic. 

The implementation with try-catch uses directly the order of statements is for flow control to increase efficiency. Specifically,  lines 264-265  log at the debug level, but (as you said) only if the bytes were successfully unpacked. If instead the bytes could not be unpacked and a `struct.error` is raised, we jump to the `except` in line 266 and, log at the error level, scan_forward and continue in line 274-276

Reply:
You can move these two lines to an `else` block belonging to the `try` block (`try`/`except`/`else`).

Reply:
```python
try:
    StreamId = struct.unpack("<I", _streamid)[0]
except struct.error:
    log_str += ", but StreamId is corrupt, scanning forward to next boundary chunk."
    logger.error(log_str)
    _scan_forward(f)
    continue
else:
    log_str += ", StreamId={}".format(StreamId)
    logger.debug(log_str)
```

Reply:
Awesome, use this so rarely, i haven't thought of it. Changed as requested.

Reply:
In theory, this isn't really necessary here (you could remove the `else` and dedent the two lines), but I think it makes the structure/intention clear.